### PR TITLE
[LAA] refactor program logic (NFC)

### PR DESCRIPTION
**Pull request by artagnon**
Implement NFC improvements spotted during a cursory reading of LoopAccessAnalysis.

### CODE REVIEW

File: llvm/include/llvm/Analysis/LoopAccessAnalysis.h [hunk 1]
```diff
@@ -540,7 +540,7 @@ class RuntimePointerChecking {
   /// Try to create add a new (pointer-difference, access size) pair to
   /// DiffCheck for checking groups \p CGI and \p CGJ. If pointer-difference
   /// checks cannot be used for the groups, set CanUseDiffCheck to false.
-  void tryToCreateDiffCheck(const RuntimeCheckingPtrGroup &CGI,
+  bool tryToCreateDiffCheck(const RuntimeCheckingPtrGroup &CGI,
```
[the remark this answers is not in the document]
Comment (artagnon) [reply llvm/include/llvm/Analysis/LoopAccessAnalysis.h]: See #92110.

**Comment by artagnon** [comment]
Sorry, I made a mess and SLPVectorizer tests are failing. Inspecting the diff now.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 2]
```diff
@@ -1646,22 +1644,19 @@ bool llvm::sortPtrAccesses(ArrayRef<Value *> VL, Type *ElemTy,
```
Comment (fhahn) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:1644]: might be good to split off the ::sortPtrAccesses changes separately, as this is what SLP uses I think

**Comment by artagnon** [comment]
All problems fixed, and sortPtrAccesses() split out into #92256.

**Comment by artagnon** [comment]
Gentle ping. We should perhaps get this out of the way, before future changes cause conflicts?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 3]
```diff
@@ -394,9 +394,9 @@ void RuntimePointerChecking::generateChecks(
 
 bool RuntimePointerChecking::needsChecking(
     const RuntimeCheckingPtrGroup &M, const RuntimeCheckingPtrGroup &N) const {
-  for (unsigned I = 0, EI = M.Members.size(); EI != I; ++I)
-    for (unsigned J = 0, EJ = N.Members.size(); EJ != J; ++J)
-      if (needsChecking(M.Members[I], N.Members[J]))
+  for (auto &I : M.Members)
```
Comment (fhahn) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:397]: nit: `const auto &`?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 4]
```diff
@@ -2082,14 +2074,12 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
   // Negative distances are not plausible dependencies.
   if (SE.isKnownNonPositive(Dist)) {
     if (SE.isKnownNonNegative(Dist)) {
-      if (HasSameSize) {
+      if (HasSameSize)
```
Comment (fhahn) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2077]: please retain braces here when the body is multiple lines (https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 5]
```diff
@@ -3050,11 +3041,10 @@ LoopAccessInfo::LoopAccessInfo(Loop *L, ScalarEvolution *SE,
   if (TTI) {
     TypeSize FixedWidth =
         TTI->getRegisterBitWidth(TargetTransformInfo::RGK_FixedWidthVector);
-    if (FixedWidth.isNonZero()) {
+    if (FixedWidth.isNonZero())
```
Comment (fhahn) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:3044]: please retain braces here when the body is multiple lines (https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

In the future, it would be good to avoid force pushes and use merge commits instead of rebases to update the PR, as Github deals with merges much better than with force pushes & rebases in terms of retaining comments

**Comment by artagnon** [comment]
> In the future, it would be good to avoid force pushes and use merge commits instead of rebases to update the PR, as Github deals with merges much better than with force pushes & rebases in terms of retaining comments

I must admit that I'm a little nervous about using merge commits, since `main` requires linear history, and I'm worried that I might mess something up. Besides, I like to view my branch history locally before submitting, and intervening merge commits make it complicated to do so.

**Comment by fhahn** [comment]
> > In the future, it would be good to avoid force pushes and use merge commits instead of rebases to update the PR, as Github deals with merges much better than with force pushes & rebases in terms of retaining comments
> 
> I must admit that I'm a little nervous about using merge commits, since `main` requires linear history, and I'm worried that I might mess something up. Besides, I like to view my branch history locally before submitting, and intervening merge commits make it complicated to do so.

When merging via GitHub, squash-and-merge is the only option, so there should be nothing to worry about in that respect.

Please see https://llvm.org/docs//GitHub.html#rebasing-pull-requests-and-force-pushes for the official recommendation. While a linear history locally might be convenient, preserving the comment and change history as much as possible is also very important for the review process. This is mostly a consequence of how Github PRs work and has been discussed at length before the switch.

**Comment by nikic** [comment]
FWIW, I don't agree with @fhahn and every time a merge commit is introduced in a PR is a bit annoyance to me. Please do prefer rebase, just avoid also squashing commits at the same time.

**Comment by fhahn** [comment]
> FWIW, I don't agree with @fhahn and every time a merge commit is introduced in a PR is a big annoyance to me. Please do prefer rebase, just avoid also squashing commits at the same time.

Hmm, looking back the link I shared, it seems the recommendation is to not update the PR at all (also no merge) unless absolutely necessary. Also looking back at the discussion, there doesn't seem to be a very clear consensus either way.

 It's a shame that GH doesn't handle this better, e.g. it is not possible the view the earlier code comments in that PR in the code view (https://github.com/llvm/llvm-project/pull/92101/files#diff-4fcef9d109d67042b3c7d8d5a005ace05d1a0ac7e6cdee8dec106352c1dd8447) but I guess  it is something we have to live with now

**Pull request merged**